### PR TITLE
Add Nested groups

### PR DESCRIPTION
**Pull request by yotamberk**
solves #2300

### CODE REVIEW

**Comment by yotamberk** [comment]
Still missing docs. Working on it...

File: lib/timeline/component/ItemSet.js [hunk 1]
```diff
@@ -25,231 +25,232 @@ var BACKGROUND = '__background__'; // reserved group id for background items wit
  * @extends Component
  */
 function ItemSet(body, options) {
-  this.body = body;
```
Comment (mojoaxel) [inline lib/timeline/component/ItemSet.js]: You updated the indent to `tabs`. I like that! Sadly this makes this Pull-Request un-reviewable. Could you maybe change that back?

Comment (yotamberk) [inline lib/timeline/component/ItemSet.js]: Can you recommend an easy way to revert this specific commit?

File: lib/timeline/component/Group.js [hunk 2]
```diff
@@ -50,12 +61,17 @@ function Group (groupId, data, itemSet) {
...
   if (this.itemSet.options.groupEditable.order) {
-	  label.className = 'vis-label draggable';
+    label.className = 'vis-label draggable';
   } else {
-	  label.className = 'vis-label';
+    label.className = 'vis-label';
   }
   this.dom.label = label;
 
+  var collapseGroupBtn = document.createElement('div');
```
Comment (mojoaxel) [inline lib/timeline/component/Group.js]: I think the whole `collapseGroupBtn`can be removed, now that the arrow is done is CSS.

File: lib/timeline/component/css/itemset.css [hunk 3]
```diff
@@ -33,6 +33,26 @@
   border-bottom: none;
 }
 
+.vis-nesting-group {
+  cursor: pointer;
+}
+
+.vis-nested-group {
+  background: #e5e5e5;
```
Comment (mojoaxel) [inline lib/timeline/component/css/itemset.css]: I personally would prefer something even a little lighter like `#f5f5f5`

File: lib/timeline/component/Group.js [hunk 4]
```diff
@@ -113,14 +129,43 @@ Group.prototype.setData = function(data) {
...
+    } else {
+      this.showNested = true;
+    }
+
+    this.dom.collapseGroupBtn.style.display = 'inline-block';
+    util.addClassName(this.dom.label, 'vis-nesting-group');
+    if (this.showNested) {
+      util.removeClassName(this.dom.label, 'collapsed');
+      util.addClassName(this.dom.label, 'expanded');
+      // this.dom.collapseGroupBtn.innerHTML = '&#9660;';
```
Comment (mojoaxel) [inline lib/timeline/component/Group.js]: This comments can be removed to keep the code clean

**Comment by yotamberk** [comment]
@mojoaxel Fixed the comments =)

File: lib/timeline/component/ItemSet.js [hunk 5]
```diff
@@ -1611,15 +1690,16 @@ ItemSet.prototype._onGroupDrag = function (event) {
 			
 			// switch groups
 			if (draggedGroup && targetGroup) {
-				this.options.groupOrderSwap(draggedGroup, targetGroup, this.groupsData);
-  				groupsData.update(draggedGroup);
-        			groupsData.update(targetGroup);
+				this.options.groupOrderSwap(draggedGroup, targetGroup, groupsData);
+				groupsData.update(draggedGroup);
+      	groupsData.update(targetGroup);
```
Comment (mojoaxel) [inline lib/timeline/component/ItemSet.js]: indent =)

File: lib/timeline/component/ItemSet.js [hunk 6]
```diff
@@ -1721,7 +1801,7 @@ ItemSet.prototype._onGroupDragEnd = function (event) {
   					var shouldBeGroup = dataset.get(origOrder[curPos]);
   					me.options.groupOrderSwap(switchGroup, shouldBeGroup, dataset);
   					groupsData.update(switchGroup);
-					groupsData.update(shouldBeGroup);
+					  groupsData.update(shouldBeGroup);
```
Comment (mojoaxel) [inline lib/timeline/component/ItemSet.js]: indent =)

File: examples/timeline/groups/nestedGroups.html [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      content: "cook",
+    },
+    {
+      id: 12,
+      content: "shop",
+    },
+    {
+      id: 13,
+      content: "clean house",
+
```
Comment (mojoaxel) [inline examples/timeline/groups/nestedGroups.html]: empty line

File: docs/timeline/index.html [hunk 8]
```diff
@@ -461,6 +461,18 @@ <h3 id="groups">Groups</h3>
       <td>no</td>
       <td>Provides a means to toggle the whether a group is displayed or not. Defaults to <code>true</code>.</td>
     </tr>
+    <tr>
+      <td>nestedGroups</td>
```
Comment (mojoaxel) [inline docs/timeline/index.html]: Now that I look at it I think "subGroups" would have been the better work, but I'm ok with this.

File: examples/timeline/groups/nestedGroups.html [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    {
+      id: 14,
+      content: "wash dishes",
+    }
+  ]);
+
+
+  // create a dataset with items
+  var items = new vis.DataSet();
+  var groupIds = groups.getIds()
```
Comment (mojoaxel) [inline examples/timeline/groups/nestedGroups.html]: missing semicolon

File: examples/timeline/groups/nestedGroups.html [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    },
+    {
+      id: 2,
+      content: "Alson",
+    },
+    {
+      id: 3,
+      content: "John",
+      nestedGroups: [14],
+      showNested: false
```
Comment (mojoaxel) [inline examples/timeline/groups/nestedGroups.html]: The `showNested: false` causes the bad error `Cannot redraw item: no parent attached`. Looks like Items can not be attached to groups that are not visible.
This needs to be fixed !

Comment (yotamberk) [inline examples/timeline/groups/nestedGroups.html]: I'm on it at this moment. It's a weird problem. trying to figure out what is causing this exactly...

File: docs/timeline/index.html (hunk 8 above)
Comment (yotamberk) [inline docs/timeline/index.html]: subGroups refer to groups in a group and not nested underneath. So I actually prefer nestedGroups.

**Comment by yotamberk** [comment]
@mojoaxel Hey! I just managed to figure out how to fix that warning that occurred with my PR. Took me 4 days but managed eventually... It seems like it was a warning that already occurred with the addition of the visible group. Trying to initialize a group with `visible:false` was also causing this issue. I fixed it for all cases.
Can you please review and merge this? I want to get rid of the PRs that are accumulating

**Comment by yotamberk** [comment]
@mojoaxel Can you please review this again and accept it?

**Pull request merged**